src/System.CommandLine.Hosting/HostingExtensions.cs

```diff
@@ -64,5 +66,31 @@ public static class HostingExtensions
                     services.Configure(configureOptions);
             });
         }
+
+        public static InvocationContext GetInvocationContext(this IHostBuilder hostBuilder) =>
+            GetInvocationContext(hostBuilder?.Properties);
+
+        public static InvocationContext GetInvocationContext(this HostBuilderContext context) =>
+            GetInvocationContext(context?.Properties);
+
+        private static InvocationContext GetInvocationContext(IDictionary<object, object> properties)
+        {
+            object ctxObj = null;
+            if (properties?.TryGetValue(typeof(InvocationContext), out ctxObj) ?? false)
+                return ctxObj as InvocationContext;
+            return null;
+        }
+
+        public static void ConfigureFromCommandLine<TOptions>(
```

Comment:
What do y'all think of `ConfigureCommand` as a name? `ConfigureFromCommandLine` is a little long

Reply:
The thing being configured is the `IServiceCollection`, not a command,  no?

Reply:
correct, or to be more precise, it's configuring the instance of `TOptions` passed as generic type argument.

Reply:
To be utterly and totally pedantic:

```
ConfigureServiceProviderToUseCommandLineParseResultToCreateInstancesOf<T>
```

Now remove the words that aren't adding critical information and you have your method name.

Reply:
I still think `ConfigureFromCommandLine` is best.

Technically `ConfigureFromBindingContext` would also be fairly accurate, but that would also be a little overloaded (similar to `ConfigureFromInvocation`).

Reply:
And the method does not actually "configure the service provider to use command line parse result to create an instance of T". The actual method signature I am basing this on is `Configure<TOptions>() where TOptions : class, new()`.

So the implementation gets an already instantiated `TOptions` and is expected to set settable properties on that type. It does so using `ModelBinder<TOptions>.UpdateInstance(opts, bindingContext)`.

That leads the DI container to know how to resolve `IOption<TOption>` not actually directly the instance of `TOption`!

To register `TOption` you'd actually call `AddSingleton`, `AddScoped` or `AddTransient`, and all those overloads already have the possibility to resolve other registered DI services, no that we could already do before. `Configure<TOptions>` however does not a have a way of resolving DI services, that's why we needed to get the `BindingContext` from somewhere else. (*`BindingContext` is already and has always been registered as a DI singleton when `UseHost` is invoked.*)

Reply:
> So the implementation gets an already instantiated TOptions and is expected to set settable properties on that type. It does so using ModelBinder<TOptions>.UpdateInstance(opts, bindingContext).

Out of curiosity, why favor `IOptions<T>` over simply `T` and therefore require setters over constructor parameters?

Reply:
> The thing being configured is the `IServiceCollection`, not a command, no?

Yes, but this lines up the rest of the ecosystem. For example, `services.Configure<T>(...)` adds a `IConfigureOptions<T>` to the collection. I don't think we should establish a new `AFromB` pattern.

> Out of curiosity, why favor IOptions<T> over simply T and therefore require setters over constructor parameters?

Also, you can setup a hierarchy of configurations. I can set some defaults from my app, let an `appsettings.json` file override the defaults, and finally let my CLI arguments do some final overrides. Silly example:

```csharp
services.Configure<CliOptions>(options =>
{
  options.A = "Default value";
});

services.Configure<CliOptions>(options =>
{
  options.A = "Override value";
});
```

This model is very powerful, especially when you have many configuration sources (defaults from app, appsettings.json for local development, Key Vault, Azure App Service settings, Azure App Configuration)

Reply:
Another example: Azure SDK for .NET uses `services.ConfigureDefaults` (see [this](https://github.com/Azure/azure-sdk-for-net/blob/e0bdf52750a60af46c62973423674118d67bb105/sdk/core/Microsoft.Extensions.Azure/samples/Startup.cs#L46)). The paradigm seems to be `ConfigureThing`, so I think `ConfigureCommand` follows that pattern best. Please let me know if you'd like more example, I can find more.

Reply:
Still, `ConfigureFromCommandLine` is still the best and most accurate name.

Reply:
okay what about this:

``` C#
.UseHost(host =>
{
  host.ConfigureServices(services =>
  {
    services.AddOptions<MyCommandLineOptions>()
      .UseCommandLine();
  });
})
```

The change is:

* Rename method to `UseCommandLine`
* Extend `OptionsBuilder<TOptions>` instead of `IServiceCollection`

Reply:
I actually like that one soo much, I am going to add it anyways 😂